From c683b1265312826abbfbc296f7f892c4e6fd2ec3 Mon Sep 17 00:00:00 2001
From: Ranjith Katravula <RanjithKumar.Katravula@amd.com>
Date: Fri, 6 Mar 2026 10:37:34 +0530
Subject: [PATCH] visp: visp_video: Add stride-aware format handling

- Extend cam_device output format metadata with:
  - buf_offset (YUV/RAW offsets)
  - buf_stride (Y/RAW stride)
- Populate IspFmt->buf_stride for YUV and RAW media pixel formats in
  media_fmt_to_isp_fmt(), including packed/aligned RAW variants.
- Update visp_video_mfmt_to_vfmt() to compute bytesperline with explicit
  alignment handling for packed YUV/RGB formats and private VISP formats.
- Align bytesperline and sizeimage consistently for single-plane and
  multi-plane formats.

This improves consistency between ISP format mapping and V4L2 buffer layout
expectations when stride alignment is required.

Signed-off-by: Ranjith Katravula <RanjithKumar.Katravula@amd.com>
---
 common/visp_align.h                           | 77 +++++++++++++++++++
 visp/visp_app.c                               | 53 +++++++++++++
 visp/visp_procfs.c                            | 36 ++++++++-
 .../include/cam_device/cam_device_api.h       |  2 +
 .../include/cam_device/cam_device_common.h    | 24 ++++++
 visp_video/visp_video_register.c              | 40 ++++++++--
 6 files changed, 225 insertions(+), 7 deletions(-)
 create mode 100644 common/visp_align.h
 mode change 100644 => 100755 visp/visp_app.c
 mode change 100644 => 100755 visp/vsi_units/include/cam_device/cam_device_api.h
 mode change 100644 => 100755 visp/vsi_units/include/cam_device/cam_device_common.h

diff --git a/common/visp_align.h b/common/visp_align.h
new file mode 100644
index 0000000..f94d059
--- /dev/null
+++ b/common/visp_align.h
@@ -0,0 +1,77 @@
+/* SPDX-License-Identifier: MIT */
+/****************************************************************************
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2025 VeriSilicon Holdings Co., Ltd.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ *****************************************************************************
+ *
+ * The GPL License (GPL)
+ *
+ * Copyright (c) 2025 VeriSilicon Holdings Co., Ltd.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program;
+ *
+ *****************************************************************************
+ *
+ * Note: This software is released under dual MIT and GPL licenses. A
+ * recipient may use this file under the terms of either the MIT license or
+ * GPL License. If you wish to use only one license not the other, you can
+ * indicate your decision by deleting one of the above license notices in your
+ * version of this file.
+ *
+ *****************************************************************************/
+#ifndef __VISP_ALIGN_H__
+#define __VISP_ALIGN_H__
+
+/*
+ * Runtime-configurable stride alignment in bytes.
+ * Both visp and visp_video modules reference this shared variable.
+ * Default: 16 bytes
+ * Can be modified via: /proc/visp-isp0/align_bytes (or similar procfs entry)
+ */
+extern unsigned int visp_stride_align_bytes;
+
+/*
+ * Runtime-configurable vb2 queue minimum queued buffers.
+ * Used by visp_video queue initialization.
+ * Default: 5
+ * Can be modified via: /proc/visp-ispX/visp-isp-subdevY using min_queued_buffers
+ */
+extern unsigned int visp_video_min_queued_buffers;
+
+#ifndef ALIGN_BYTES
+#define ALIGN_BYTES (visp_stride_align_bytes)
+#endif
+
+#endif /* __VISP_ALIGN_H__ */
diff --git a/visp/visp_app.c b/visp/visp_app.c
old mode 100644
new mode 100755
index 9eb2e8b..8359a79
--- a/visp/visp_app.c
+++ b/visp/visp_app.c
@@ -71,11 +71,20 @@
 #include <linux/string.h>
 #include <media/v4l2-device.h>
 #include <media/v4l2-event.h>
+#include "visp_align.h"
 
 #define MEDIA_ISP_MCM_BUF_COUNT 4
 #define MEDIA_ISP_EMPTY_BUF_WAIT_TIME 10
 #define MEDIA_ISP_FULL_BUF_WAIT_TIME 120
 
+/* Shared runtime stride alignment for all VISP modules */
+unsigned int visp_stride_align_bytes = 16;
+EXPORT_SYMBOL_GPL(visp_stride_align_bytes);
+
+/* Shared runtime vb2 minimum queued buffers for VISP video queue */
+unsigned int visp_video_min_queued_buffers = 5;
+EXPORT_SYMBOL_GPL(visp_video_min_queued_buffers);
+
 int visp_set_fmt_public(struct visp_dev *isp_dev,
 			struct v4l2_subdev_format *format);
 
@@ -1096,98 +1105,127 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_YUYV:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422I;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_NV16:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422SP;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_NV12:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV420SP;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_GREY:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV400;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_Y10BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV400;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 10, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_Y10DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV400_ALIGNED_MODE0;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width * 4 / 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_Y10:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV400_ALIGNED_MODE1;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 8) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P00BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV420SP;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 10, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P00DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV420SP_ALIGNED_MODE0;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width * 4 / 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P010:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV420SP_ALIGNED_MODE1;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 8) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P02BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV420SP;
 		IspFmt->data_bits = 12;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 12, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P20BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422SP;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 10, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P20DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422SP_ALIGNED_MODE0;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width * 4 / 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P210:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422SP_ALIGNED_MODE1;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 8) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_P22BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422SP;
 		IspFmt->data_bits = 12;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 12, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_I20BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422I;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 10, 128) * 16, ALIGN_BYTES);
+		//IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 10, 128) * 16 * 4, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_I210:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV422I_ALIGNED_MODE1;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 8) * 16, ALIGN_BYTES);
+		//IspFmt->buf_stride.y = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 8) * 16 * 2, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_M48BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV444P;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_I48BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV444I;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
+		//IspFmt->buf_stride.y = ALIGN(IspFmt->out_width * 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_I48DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV444I_ALIGNED_MODE0;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(((IspFmt->out_width * 4) / 3) * 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_I40DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_YUV444I_ALIGNED_MODE0;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.y = ALIGN(((IspFmt->out_width * 4) / 3), ALIGN_BYTES);
+		//IspFmt->buf_stride.y = ALIGN(((IspFmt->out_width * 4) / 3) * 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_RGB24:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RGB888;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
+		//IspFmt->buf_stride.y = ALIGN(IspFmt->out_width * 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_RGB24DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RGB888_ALIGNED_MODE0;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN((IspFmt->out_width * 4 / 3) * 3, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_RGB24P:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RGB888P;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.y = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR8:
 	case MEDIA_PIX_FMT_SGBRG8:
@@ -1195,6 +1233,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB8:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW8;
 		IspFmt->data_bits = 8;
+		IspFmt->buf_stride.raw = ALIGN(IspFmt->out_width, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR10:
 	case MEDIA_PIX_FMT_SGBRG10:
@@ -1202,6 +1241,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB10:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW10_ALIGNED_MODE1;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.raw = ALIGN(IspFmt->out_width * 2, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR10BPACK:
 	case MEDIA_PIX_FMT_SGBRG10BPACK:
@@ -1209,6 +1249,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB10BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW10;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width  * 10, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR10DWA:
 	case MEDIA_PIX_FMT_SGBRG10DWA:
@@ -1216,6 +1257,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB10DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW10_ALIGNED_MODE0;
 		IspFmt->data_bits = 10;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 12) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR12:
 	case MEDIA_PIX_FMT_SGBRG12:
@@ -1223,6 +1265,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB12:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW12_ALIGNED_MODE1;
 		IspFmt->data_bits = 12;
+		IspFmt->buf_stride.raw = ALIGN(IspFmt->out_width * 2, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR12BPACK:
 	case MEDIA_PIX_FMT_SGBRG12BPACK:
@@ -1230,6 +1273,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB12BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW12;
 		IspFmt->data_bits = 12;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width  * 12, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR12DWA:
 	case MEDIA_PIX_FMT_SGBRG12DWA:
@@ -1237,6 +1281,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB12DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW12_ALIGNED_MODE0;
 		IspFmt->data_bits = 12;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 10) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR14:
 	case MEDIA_PIX_FMT_SGBRG14:
@@ -1244,6 +1289,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB14:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW14_ALIGNED_MODE1;
 		IspFmt->data_bits = 14;
+		IspFmt->buf_stride.raw = ALIGN(IspFmt->out_width * 2, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR14BPACK:
 	case MEDIA_PIX_FMT_SGBRG14BPACK:
@@ -1251,6 +1297,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB14BPACK:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW14;
 		IspFmt->data_bits = 14;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width  * 14, 128) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR14DWA:
 	case MEDIA_PIX_FMT_SGBRG14DWA:
@@ -1258,6 +1305,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB14DWA:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW14_ALIGNED_MODE0;
 		IspFmt->data_bits = 14;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width, 9) * 16, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR16:
 	case MEDIA_PIX_FMT_SGBRG16:
@@ -1265,6 +1313,7 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB16:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW16;
 		IspFmt->data_bits = 16;
+		IspFmt->buf_stride.raw = ALIGN(IspFmt->out_width * 2, ALIGN_BYTES);
 		break;
 	case MEDIA_PIX_FMT_SBGGR24:
 	case MEDIA_PIX_FMT_SGBRG24:
@@ -1272,11 +1321,15 @@ static int media_fmt_to_isp_fmt(uint32_t *media_fmt,
 	case MEDIA_PIX_FMT_SRGGB24:
 		IspFmt->out_format = CAMDEV_PIX_FMT_RAW24;
 		IspFmt->data_bits = 24;
+		IspFmt->buf_stride.raw = ALIGN(DIV_ROUND_UP(IspFmt->out_width * 24, 128) * 16, ALIGN_BYTES);
 		break;
 	default:
 		printk(KERN_ERR "Not support format %s", (char *)media_fmt);
 		return VSI_ERR_NOT_SUPPORT;
 	}
+
+	printk("%s format:%x stride.y:%d stride.raw:%d\n", __func__, *media_fmt,
+		IspFmt->buf_stride.y, IspFmt->buf_stride.raw);
 	return VSI_SUCCESS;
 }
 
diff --git a/visp/visp_procfs.c b/visp/visp_procfs.c
index 3f8ee0d..331845c 100644
--- a/visp/visp_procfs.c
+++ b/visp/visp_procfs.c
@@ -60,6 +60,7 @@
 #include <linux/version.h>
 #include "visp_driver.h"
 #include "visp_procfs.h"
+#include "visp_align.h"
 
 struct visp_procfs {
 	struct proc_dir_entry *pde;
@@ -105,6 +106,10 @@ static int visp_procfs_info_show(struct seq_file *sfile, void *offset)
 		if (pad && (port != (pad_idx / VISP_PORT_PAD_NR))) {
 			port = pad_idx / VISP_PORT_PAD_NR;
 			seq_printf(sfile, "isp%d port%d:\n", isp_dev->id, port);
+			seq_printf(sfile, "align_bytes : %u\n",
+				visp_stride_align_bytes);
+			seq_printf(sfile, "min_queued_buffers : %u\n",
+				visp_video_min_queued_buffers);
 			seq_printf(sfile, "output_type: %s %s %s %s\n",
 				   type2str(isp_dev->output_type[port][0]),
 				   type2str(isp_dev->output_type[port][1]),
@@ -261,8 +266,37 @@ static int32_t visp_proc_process(struct seq_file *sfile,
 						       .sensor_info.one_json));
 					strncpy(isp_dev->isp_ports[port]
 						    .sensor_info.one_json,
-						val, strlen(val));
+						val, strlen(val));			}
+		} else if (strcmp(val, "align_bytes") == 0) {
+			val = strsep(&kv_cur, kv_delim);
+			if (val && isdigit(*val)) {
+				unsigned int new_align =
+					(unsigned int)simple_strtoul(val, &end, 0);
+				if (new_align > 0 && new_align <= 256) {
+					visp_stride_align_bytes = new_align;
+					seq_printf(sfile, "align_bytes set to %u\n",
+						visp_stride_align_bytes);
+				} else {
+					seq_printf(sfile,
+						"Invalid align_bytes value (must be 1-256)\n");
+				}				}
+		} else if (strcmp(val, "min_queued_buffers") == 0) {
+			val = strsep(&kv_cur, kv_delim);
+			if (val && isdigit(*val)) {
+				unsigned int new_min_queued_buffers =
+					(unsigned int)simple_strtoul(val, &end, 0);
+				if (new_min_queued_buffers > 0 &&
+				    new_min_queued_buffers <= 32) {
+					visp_video_min_queued_buffers =
+						new_min_queued_buffers;
+					seq_printf(sfile,
+						"min_queued_buffers set to %u\n",
+						visp_video_min_queued_buffers);
+				} else {
+					seq_printf(sfile,
+						"Invalid min_queued_buffers value (must be 1-32)\n");
 				}
+			}
 			} else if (strcmp(val, "output_type") == 0) {
 				val = strsep(&kv_cur, kv_delim);
 				if (val) {
diff --git a/visp/vsi_units/include/cam_device/cam_device_api.h b/visp/vsi_units/include/cam_device/cam_device_api.h
old mode 100644
new mode 100755
index 5b280ff..0042cef
--- a/visp/vsi_units/include/cam_device/cam_device_api.h
+++ b/visp/vsi_units/include/cam_device/cam_device_api.h
@@ -196,6 +196,8 @@ typedef struct cam_device_pipe_out_fmt_s {
 						three yuv channal save in ddr */
 	cam_device_mi_swap_u
 	    swap; /**< Output format swap control information */
+	cam_device_mi_buf_offset_t buf_offset;
+	cam_device_mi_buf_stride_t buf_stride;
 } cam_device_pipe_out_fmt_t;
 
 /*****************************************************************************/
diff --git a/visp/vsi_units/include/cam_device/cam_device_common.h b/visp/vsi_units/include/cam_device/cam_device_common.h
old mode 100644
new mode 100755
index 8254425..e34821f
--- a/visp/vsi_units/include/cam_device/cam_device_common.h
+++ b/visp/vsi_units/include/cam_device/cam_device_common.h
@@ -221,6 +221,30 @@ typedef union cam_device_mi_swap_s {
 	cam_device_mi_swap_type_t raw_swap; /**< Swap type of RAW format. */
 } cam_device_mi_swap_u;
 
+/*****************************************************************************/
+/**
+ * @brief   structure type to specify the Raw and Yuv buffer offset
+ *
+ *****************************************************************************/
+typedef struct cam_device_mi_buf_offset_s {
+    struct {
+        uint32_t y;    /**< offset of Y channel. */
+        uint32_t u;    /**< offset of U channel. */
+        uint32_t v;    /**< offset of V channel. */
+    } yuv_offset;    /**< The offset of YUV. */
+    uint32_t raw_offset;    /**< offset of RAW. */
+} cam_device_mi_buf_offset_t;
+
+/*****************************************************************************/
+/**
+ * @brief   structure type to specify the Raw and Yuv buffer stride
+ *
+ *****************************************************************************/
+typedef struct cam_device_mi_buf_stride_s {
+    uint32_t y;    /**< offset of RAW. */
+    uint32_t raw;    /**< offset of RAW. */
+} cam_device_mi_buf_stride_t;
+
 /*****************************************************************************/
 /**
  * @brief   Enumeration type to specify the order of YUV or rgb channel
diff --git a/visp_video/visp_video_register.c b/visp_video/visp_video_register.c
index 2c0576a..1278933 100644
--- a/visp_video/visp_video_register.c
+++ b/visp_video/visp_video_register.c
@@ -71,7 +71,7 @@
 #include <media/v4l2-ioctl.h>
 #include <media/v4l2-mc.h>
 #include <media/videobuf2-dma-contig.h>
-
+#include "visp_align.h"
 #include "visp_v4l2_common.h"
 #include "visp_v4l2_std_exts.h"
 
@@ -1263,7 +1263,23 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 	height = f->fmt.pix.height;
 	info = v4l2_format_info(f->fmt.pix.pixelformat);
 	if (info != NULL) {
-		bytesperline = info->bpp[0] * width;
+		switch (info->format) {
+		case V4L2_PIX_FMT_YUYV:
+		case V4L2_PIX_FMT_YVYU:
+		case V4L2_PIX_FMT_UYVY:
+		case V4L2_PIX_FMT_VYUY:
+			bytesperline = ALIGN(width, ALIGN_BYTES) * 2;
+			break;
+
+		case V4L2_PIX_FMT_RGB24:
+		case V4L2_PIX_FMT_BGR24:
+			bytesperline = ALIGN(width, ALIGN_BYTES) * 3;
+			break;
+
+		default:
+			bytesperline = info->bpp[0] * width;
+			break;
+		}
 	} else {
 		private_fmt = 1;
 		info = visp_video_vfmt_info(f->fmt.pix.pixelformat);
@@ -1278,8 +1294,10 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 		case V4L2_PIX_FMT_M48BPACK:
 		case V4L2_PIX_FMT_I48BPACK:
 			bytesperline = width;
-			if (info->format == V4L2_PIX_FMT_I48BPACK)
+			if (info->format == V4L2_PIX_FMT_I48BPACK) {
+				bytesperline = ALIGN(bytesperline, ALIGN_BYTES);
 				bytesperline *= 3;
+			}
 			break;
 
 		// 8bit dwa
@@ -1294,8 +1312,10 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 		case V4L2_PIX_FMT_P20BPACK:
 		case V4L2_PIX_FMT_I20BPACK:
 			bytesperline = DIV_ROUND_UP(width * 10, 128) * 16;
-			if (info->format == V4L2_PIX_FMT_I20BPACK)
+			if (info->format == V4L2_PIX_FMT_I20BPACK) {
+				bytesperline = ALIGN(bytesperline, ALIGN_BYTES);
 				bytesperline *= 4;
+			}
 			break;
 
 		// 10 bit double align
@@ -1311,8 +1331,10 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 		case V4L2_PIX_FMT_P210:
 		case V4L2_PIX_FMT_I210:
 			bytesperline = DIV_ROUND_UP(width, 8) * 16;
-			if (info->format == V4L2_PIX_FMT_I210)
+			if (info->format == V4L2_PIX_FMT_I210) {
+				bytesperline = ALIGN(bytesperline, ALIGN_BYTES);
 				bytesperline *= 2;
+			}
 			break;
 
 		// 12bit unalign
@@ -1406,8 +1428,9 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 		}
 	}
 
+	bytesperline = ALIGN(bytesperline, ALIGN_BYTES);
 	sizeimage = bytesperline * height;
-
+	sizeimage = ALIGN(sizeimage, ALIGN_BYTES);
 	if (info->comp_planes == 1) {
 		f->fmt.pix.bytesperline = bytesperline;
 		f->fmt.pix.sizeimage = sizeimage;
@@ -1424,7 +1447,9 @@ static int visp_video_mfmt_to_vfmt(struct v4l2_subdev_format *mfmt,
 			bytesperline =
 			    info->bpp[i] * DIV_ROUND_UP(width, info->hdiv);
 		}
+		bytesperline = ALIGN(bytesperline, ALIGN_BYTES);
 		sizeimage = bytesperline * DIV_ROUND_UP(height, info->vdiv);
+		sizeimage = ALIGN(sizeimage, ALIGN_BYTES);
 		f->fmt.pix.sizeimage += sizeimage;
 	}
 
@@ -2360,6 +2385,9 @@ static int visp_video_queue_init(struct visp_video_dev *visp_vdev)
 	queue->timestamp_flags = V4L2_BUF_FLAG_TIMESTAMP_MONOTONIC;
 	queue->lock = &visp_vdev->video_lock;
 	queue->dev = visp_vdev->visp_mdev->dev;
+
+	queue->min_queued_buffers = visp_video_min_queued_buffers;
+
 	ret = vb2_queue_init(queue);
 	if (ret) {
 		dev_err(visp_vdev->visp_mdev->dev, "vb2 queue init failed\n");
-- 
2.34.1

